pkg/controller/backup/backup_controller.go

```diff
@@ -252,3 +264,191 @@ func (c *Controller) enqueueBackup(obj interface{}) {
+	name := backup.GetName()
+	jobName := backup.GetBackupJobName()
+
+	// check pod status
+	selector, err := label.NewBackup().Instance(backup.GetInstanceName()).BackupJob().Backup(name).Selector()
+	if err != nil {
+		klog.Errorf("Fail to generate selector for backup %s/%s, error %v", ns, name, err)
+		return false, "", err
+	}
+	pods, err := c.deps.PodLister.Pods(ns).List(selector)
+	if err != nil && !errors.IsNotFound(err) {
+		klog.Errorf("Fail to list pod for backup %s/%s with selector %s, error %v", ns, name, selector, err)
+		return false, "", err
+	}
+	for _, pod := range pods {
+		if pod.Status.Phase == corev1.PodFailed {
+			klog.Infof("backup %s/%s has failed pod %s.", ns, name, pod.Name)
+			return true, fmt.Sprintf("Pod %s has failed", pod.Name), nil
+		}
+	}
+
+	// to avoid missing pod failed events, double check job status
+	job, err := c.deps.JobLister.Jobs(ns).Get(jobName)
+	if err != nil && !errors.IsNotFound(err) {
+		klog.Errorf("Fail to get job %s for backup %s/%s, error %v ", jobName, ns, name, err)
+		return false, "", err
+	}
+	if job != nil {
+		for _, condition := range job.Status.Conditions {
+			if condition.Type == batchv1.JobFailed && condition.Status == corev1.ConditionTrue {
+				klog.Infof("backup %s/%s has failed job %s", ns, name, jobName)
+				return true, fmt.Sprintf("Job %s has failed", jobName), nil
+			}
+		}
+	}
+	return false, "", nil
+}
+
+func (c *Controller) tryToRestartSnapshotBackup(backup *v1alpha1.Backup, reason string) error {
+	var (
+		ns            = backup.GetNamespace()
+		name          = backup.GetName()
+		isTimeToRetry bool
+		err           error
+	)
+	klog.V(4).Infof("try to restart backup %s/%s, failed reason %s", ns, name, reason)
+
+	if c.isExceedRetryLimit(backup) {
+		klog.Errorf("backup %s/%s exceed retry limit, failed reason %s", ns, name, reason)
+		err = c.control.UpdateCondition(backup, &v1alpha1.BackupCondition{
+			Type:    v1alpha1.BackupFailed,
+			Status:  corev1.ConditionTrue,
+			Reason:  "AlreadyFailed",
+			Message: fmt.Sprintf("exceed retry limit, failed reason %s", reason),
+		})
+		if err != nil {
+			klog.Errorf("Fail to update the condition of backup %s/%s, %v", ns, name, err)
+			return err
+		}
+		return nil
+	}
+
+	isTimeToRetry, err = c.isTimeToRetry(backup)
+	if err != nil {
+		klog.Errorf("Fail to check whether is the time to retry of backup %s/%s", ns, name)
+		return err
+	}
+
+	if !isTimeToRetry {
+		klog.V(4).Infof("backup %s/%s is not the time to retry, next retry time is %s", ns, name, backup.Status.ExponentialBackoffRetry.NextRetryAt)
+		return nil
+	}
+
+	err = c.cleanBackupOldJob(backup)
+	if err != nil {
+		klog.Errorf("Fail to clean job of backup %s/%s, error is %v", ns, name, err)
+		return err
+	}
+
+	err = c.resetRetryRecordToNext(backup)
+	if err != nil {
+		klog.Errorf("Fail to refresh retry record of backup %s/%s, error is %v", ns, name, err)
+		return err
+	}
+
+	// retry
+	err = c.control.UpdateCondition(backup, &v1alpha1.BackupCondition{
```

Comment:
Why update status to `retryFailed` and `restart` in succession? Why not just set it `restart`?